test/helpers/noisydgrambio.c

```diff
@@ -0,0 +1,290 @@
+    struct noisy_dgram_st *data;
+
+    if (!TEST_ptr(next))
+        return 0;
+
+    data = BIO_get_data(bio);
+    if (!TEST_ptr(data))
+        return 0;
+
+    /*
+     * For simplicity we assume that all elements in the msg array have the
+     * same data_len. They are not required to by the API, but it would be quite
+     * strange for that not to be the case - and our code that calls
+     * BIO_recvmmsg does do this (which is all that is important for this test
+     * code). We test the invariant here.
+     */
+    for (i = 0; i < num_msg; i++) {
+        if (i == 0) {
+            data_len = msg[i].data_len;
+            if (!TEST_size_t_le(data_len, MSG_DATA_LEN_MAX))
+                return 0;
+        } else if (!TEST_size_t_eq(msg[i].data_len, data_len)) {
+            return 0;
+        }
+    }
+
+    if (!BIO_recvmmsg(next, msg, stride, num_msg, flags, msgs_processed))
+        return 0;
+
+#ifdef OSSL_NOISY_DGRAM_DEBUG
+    printf("Pre-filter datagram list:\n");
+    for (i = 0; i < *msgs_processed; i++) {
+        printf("Pre-filter Datagram:\n");
+        BIO_dump_fp(stdout, msg[i].data, msg[i].data_len);
+        printf("\n");
+    }
+    printf("End of pre-filter datagram list\nApplying noise filters:\n");
+#endif
+
+    msg_cnt = *msgs_processed;
+
+    /* Introduce noise */
+    for (i = 0, thismsg = msg;
+         i < msg_cnt;
+         i++, thismsg++, data->this_dgram++) {
+        uint64_t delay;
+        int should_drop;
+
+        /* If we have a delayed message ready insert it now */
+        if (data->delayed_dgram > 0
+                && data->delayed_dgram == data->this_dgram) {
+            if (msg_cnt < num_msg) {
+                /* Make space for the inserted message */
+                for (j = msg_cnt; j > i; j--) {
+                    if (!bio_msg_copy(&msg[j], &msg[j - 1]))
+                        return 0;
+                }
+                if (!bio_msg_copy(thismsg, &data->msg))
+                    return 0;
+                msg_cnt++;
+                data->delayed_dgram = 0;
+#ifdef OSSL_NOISY_DGRAM_DEBUG
+                printf("**Inserting a delayed datagram\n");
+                BIO_dump_fp(stdout, thismsg->data, thismsg->data_len);
+                printf("\n");
+#endif
+                continue;
+            } /* else we have no space for the insertion, so just drop it */
+            data->delayed_dgram = 0;
+        }
+
+        get_noise(&delay, &should_drop);
+
+        /* We ignore delay if a message is already delayed */
+        if (delay > 0 && data->delayed_dgram == 0) {
+            /*
+             * Note that a message may be delayed *and* dropped, or delayed
+             * and *not* dropped.
+             * Delayed and dropped means the message will not be sent now and
+             * will only be sent after the delay.
+             * Delayed and not dropped means the message will be sent now and
+             * a duplicate will also be sent after the delay.
```

Comment:
Maybe this comment should be up around line 70 when the terms are introduced for the first time.

I'm not a fan of the nomenclature used: _dropped_, _delayed_, _duplicated_ would be more in line with that is being done.

Reply:
I agree its quite confusing. I've completely revised the terminology used to try and make this clearer.